src/pykeen/training/callbacks.py

```diff
@@ -369,9 +379,102 @@ def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # n
+
+    # docstr-coverage: inherited
+    def post_batch(self, epoch: int, batch, **kwargs: Any) -> None:  # noqa: D102
+        # when called by batch_size_search(), the parameter update should not be applied.
+        if self.step:
+            # TODO: callback pre_step
+            # update parameters according to optimizer
+            self.optimizer.step()
+
+        # After changing applying the gradients to the embeddings, the model is notified that the forward
+        # constraints are no longer applied
+        self.model.post_parameter_update()
+
+
+class LearningRateSchedulerTrainingCallback(TrainingCallback):
+    """Update learning rate scheduler."""
+
+    # docstr-coverage: inherited
+    def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # noqa: D102
+        if self.training_loop.lr_scheduler is not None:
+            self.training_loop.lr_scheduler.step(epoch=epoch)
+
+
+@maximize_memory_utilization(keys=("training_loop", "id_triples_factory", "batch_size"))
+@torch.inference_mode()
+def _validation_loss_amo_wrapper(
+    training_loop: "TrainingLoop",
+    triples_factory: CoreTriplesFactory,
+    batch_size: int,
+    label_smoothing: float,
+    id_triples_factory: int,  # TODO: only used for hashing (cf. 'keys' above)
+    **kwargs,
+) -> float:
+    """Calculate validation loss with automatic batch size optimization."""
+    return training_loop._train_epoch(
+        batches=training_loop._create_training_data_loader(
+            triples_factory=triples_factory, batch_size=batch_size, drop_last=False, **kwargs
+        ),
+        label_smoothing=label_smoothing,
+        # no callbacks, no epoch
+        callbacks=MultiTrainingCallback(),
+        epoch=-1,
+        # no sub-batching (for evaluation, we can just reduce batch size without any effect)
+        sub_batch_size=None,
+        # TODO: optimize slice size?
+        slice_size=None,
+        # this is handled by the AMO wrapper
+        only_size_probing=False,
+        # no backward passes
+        backward=False,
+    )
+
+
+class ValidationLossTrainingCallback(TrainingCallback):
+    """Calculate loss on a development set."""
+
+    def __init__(self, triples_factory: CoreTriplesFactory, prefix: str = "validation"):
+        super().__init__()
+        self.triples_factory = triples_factory
+        self.prefix = prefix
+
+    def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:
+        # TODO: where to get these from?
+        label_smoothing = 0.0
+        training_data_loader_kwargs = dict(sampler=None)
+        # set to evaluation mode
+        self.model.eval()
+        loss = _validation_loss_amo_wrapper(
+            training_loop=self.training_loop,
+            triples_factory=self.triples_factory,
+            # TODO: this should be num_instances rather than num_triples; also for cpu, we may want to reduce this
+            batch_size=self.triples_factory.num_triples,
```

Comment:
This may cause OOM kills on cpu for large datasets. It would be better to derive something from the training batch size (e.g., `4*batch_size` or similar) as upper bound.

Performance-wise, a too large initial value will only effect runtime of the first call, since for later calls the previous values will be re-used.

Reply:
cf. 7d95654